src/paint/map_element/scenery.c

```diff
@@ -16,8 +16,361 @@
 
 #include "map_element.h"
 #include "../../addresses.h"
+#include "../../game.h"
+#include "../../interface/viewport.h"
+#include "../../paint/paint.h"
+#include "../../paint/supports.h"
 #include "../../world/map.h"
+#include "../../world/scenery.h"
 
 void scenery_paint(uint8 direction, int height, rct_map_element* mapElement) {
-	RCT2_CALLPROC_X(0x6DFF47, 0, 0, direction, height, (int)mapElement, 0, 0);
+	//RCT2_CALLPROC_X(0x6DFF47, 0, 0, direction, height, (int)mapElement, 0, 0); return;
+	RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8_t) = VIEWPORT_INTERACTION_ITEM_SCENERY;
+	RCT2_GLOBAL(0x009DEA52, sint16) = 0;
+	RCT2_GLOBAL(0x009DEA54, sint16) = 0;
+	int ebx = 0;
+	if (RCT2_GLOBAL(0x009DEA6F, uint8) & 1) {
+		rct_map_element *mapElement2 = RCT2_GLOBAL(0x00F63674, rct_map_element*);
+		while (mapElement2 != mapElement) {
+			mapElement2++;
+			if ((mapElement2 - 1) != (rct_map_element*)0xFFFFFFFF) {
+				continue;
+			}
+			ebx = 0x21700000;
+			break;
+		}
+	}
+	if (mapElement->flags & MAP_ELEMENT_FLAG_GHOST) {
+		RCT2_GLOBAL(RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE, uint8_t) = VIEWPORT_INTERACTION_ITEM_NONE;
+		ebx = RCT2_ADDRESS(0x993CC4, uint32_t)[RCT2_GLOBAL(RCT2_ADDRESS_CONFIG_CONSTRUCTION_MARKER, uint8)];
+	}
+	uint32 dword_F64EB0 = ebx;
+	
+	rct_scenery_entry *entry = (rct_scenery_entry*)gSmallSceneryEntries[mapElement->properties.scenery.type];;
+	ebx = entry->image + direction;
+	RCT2_GLOBAL(0x00F64ED6, uint16) = 2;
+	RCT2_GLOBAL(0x00F64ED8, uint16) = 2;
+	int ecx;
+	sint8 x_offset = 0;
+	sint8 y_offset = 0;
+	if (entry->small_scenery.flags & SMALL_SCENERY_FLAG_FULL_TILE) {
+		if (entry->small_scenery.flags & SMALL_SCENERY_FLAG24) {
+			// 6DFFE3:
+			RCT2_GLOBAL(0x009DEA52, sint16) = RCT2_ADDRESS(0x009A3E7C, uint16)[direction * 4];
+			RCT2_GLOBAL(0x009DEA54, sint16) = RCT2_ADDRESS(0x009A3E7E, uint16)[direction * 4];
+			RCT2_GLOBAL(0x00F64ED6, uint16) = RCT2_ADDRESS(0x009A3E8C, uint16)[direction * 4];
+			RCT2_GLOBAL(0x00F64ED8, uint16) = RCT2_ADDRESS(0x009A3E8E, uint16)[direction * 4];
+			x_offset = 3;
+			y_offset = 3;
+		} else {
+			x_offset = 15;
+			y_offset = 15;
+			if (entry->small_scenery.flags & SMALL_SCENERY_FLAG_VOFFSET_CENTRE) {
+				x_offset = 3;
+				y_offset = 3;
+				RCT2_GLOBAL(0x00F64ED6, uint16) = 26;
+				RCT2_GLOBAL(0x00F64ED8, uint16) = 26;
+				if (entry->small_scenery.flags & SMALL_SCENERY_FLAG_ALLOW_WALLS) {
+					x_offset = 1;
+					y_offset = 1;
+					RCT2_GLOBAL(0x00F64ED6, uint16) = 30;
+					RCT2_GLOBAL(0x00F64ED8, uint16) = 30;
+				}
+			}
+		}
+	} else {
+		// 6DFFC2:
+		ecx = ((mapElement->type >> 6) + get_current_rotation()) & 3;
+		x_offset = RCT2_ADDRESS(0x009A3E74, sint8)[ecx * 2];
+		y_offset = RCT2_ADDRESS(0x009A3E75, sint8)[ecx * 2];
+	}
+	// 6E0074:
+	RCT2_GLOBAL(0x009DEA52, sint16) = x_offset;
+	RCT2_GLOBAL(0x009DEA54, sint16) = y_offset;
+	RCT2_GLOBAL(0x009DEA56, sint16) = height;
```

Comment:
As these are now just parameters to sub_98197C etc. You can remove all references to them and pass the local variables you have.